### PR TITLE
Feature: Filter by set availability

### PR DESCRIPTION
**Pull request by grantglidewell**
Before I move forward with adding the filtering abilities I wanted to make sure the path Im on is something you would want to implement.

image

This still needs to filter the store and have a fetch variable added.

### CODE REVIEW

**Comment by speterson-riskalyze** [comment]
Obviously not the owner, but definitely feel like this is needed so +1 from me. 

I would say though that unlike OPs radio-style selection for the keycap types, this one is probably best done multi-select.

**Comment by noudadrichem** [comment]
Hee, you're ahead of me! Definitely a feature I want to implement, I need to implement more filters... Not yet sure about the design tbh, but if the logic is built I can always edit that!?.

**Comment by noudadrichem** [comment]
Could you also try to not re-format all the code and replace all single quotes with double?😳

**Comment by QuadDepo** [comment]
@grantglidewell I think the total count is not updating as shown in your picture. 

I also agree with @speterson-riskalyze and @noudadrichem that the current way we have filter might not be the best option after we added multiple profiles.

**Comment by grantglidewell** [comment]
@noudadrichem Sorry for the format change, somehow prettier formatting is broken for me with this project. Ill fix it.  What do you think about adding a `filter` object into state that encompasses any option that we add down the line to filter the view.

@QuadDepo The image is not an accurate representation of where the branch is currently, sorry.

**Comment by QuadDepo** [comment]
Alright, 👍  just something that caught my eye ;)

**Comment by grantglidewell** [comment]
> Obviously not the owner, but definitely feel like this is needed so +1 from me.
> 
> I would say though that unlike OPs radio-style selection for the keycap types, this one is probably best done multi-select.

interesting, didnt think about cross filtering with these. That adds some complexity to how I wanted to filter but totally doable.

**Comment by noudadrichem** [comment]
A filter object would be great! Some keys you could also filter on after the next release is:
- designerName
- type (is profile)
- brand
- material

as these are already in the database!

**Comment by grantglidewell** [comment]
> A filter object would be great! Some keys you could also filter on after the next release is:
> 
>     * designerName
> 
>     * type (is profile)
> 
>     * brand
> 
>     * material
> 
> 
> as these are already in the database!

sounds good, Ill take a stab at that later, in the meantime, I think maybe adding a prettier config might help as the quote formatting seems inconsistent. I see the `.editorconfig` but prettier isnt implemented in this project. 

 I think this would be a good time to use it as I think you might have more contributors and establishing and enforcing these standards on commit would prevent people from making bonehead mistakes as I have here.

**Comment by noudadrichem** [comment]
Great suggestion, tyvm. I'll add this soon!

**Comment by grantglidewell** [comment]
@noudadrichem I have the filter working on the collection of keysets that is currently in the state. Though I see something happening that we may want to change. With each filter change there is a new fetch that happens and replaces the state (at least as far as I can tell). This works, but is inefficient when you start working with a lot of filters and allowing for user entered data (search terms etc). This also has some knock on effects in the way that scroll loading happens. What I might suggest as a path forward is that the state contains as many keysets as possible (or as many that might be displayed) and they are only re-fetched when we run out of keysets to display. The filters are applied to the whole unfiltered set of keysets.

EX: a user comes to the page, no filters are applied by default, `all` sets are loaded. they select a filter, the displayed sets are immediately filtered because this happens in memory, they scroll (maybe we use [intersection observer](https://gist.github.com/grantglidewell/a3d4dc1dd30fe07f9125705e7f4e3547) on the last element to know when to load more keysets into the dom?), we run out of sets to display, and then we make a fetch call. 

We could modify the store since it is always aware of the filters to have two sets of 'keysets', `allKeysets` and `filteredKeysets`, this way we can pull that logic out of the component level. 

there are possible added benefits here where we could make fetches before the user gets to the bottom of the page. 

This change is out of scope for the changes I was trying to make so I wanted to make sure you're on board with this approach and see how we want to break this up. I hope this makes sense and please, if it doesn't, I'm happy to discuss further.

**Comment by noudadrichem** [comment]
@grantglidewell valid point about the fetching for new sets on filter. Your idea sounds very nice as it is also improving the speed of the site. So I'm in with this approach. 
Is filterForAvailability ready? Or do you want to enhance the fetching to also implement this and make it a combined thing?

**Comment by grantglidewell** [comment]
@noudadrichem I Think we can probably merge this and then make another PR that implements the other changes I laid out. Ill merge master into this branch and click the 'ready for review' button.

File: components/Images.tsx [hunk 1]
```diff
@@ -1,31 +1,58 @@
...
+        case 'waiting':
+            return moment().diff(groupbuyStartDate, 'days') < 0 && !isInterestCheck;
+        case 'gb':
+            return getDayDifference(groupbuyEndDate) > 0 && !isInterestCheck;
+        case 'ended':
+            return !isInterestCheck 
+                && !(getDayDifference(groupbuyEndDate) > 0) 
+                &&  !(moment().diff(groupbuyStartDate, 'days') < 0) 
+        default:
+            return false
```
Comment (noudadrichem) [inline components/Images.tsx:41]: I'm not sure if i like non 'CONStanized' switch cases. Maybe it's nice to add these filters to the `constants.tsx` file?

File: components/Images.tsx [hunk 2]
```diff
@@ -1,31 +1,58 @@
...
             <div className="images-container">
                 { keycapsets.length > 0 ?
                     (
                     keycapsets
                         .filter((keycapset: Keycapset) => {
-                            if (activeTab === 'all') return true;
-                            return keycapset.type === activeTab;
+                            if (activeTab === 'all' && availabilityFilter === 'none') return true;
+                            if(activeTab === 'all') return filterByAvailability(keycapset);
+                            return keycapset.type === activeTab && filterByAvailability(keycapset);
```
Comment (noudadrichem) [inline components/Images.tsx:55]: This is a lot of logic for inside the template, maybe it's nicer to make a seperate function for this to explain more what it does by e.g. the name?

File: components/Tabs/Tab.tsx [hunk 3]
```diff
@@ -1,32 +1,43 @@
...
-                        </Button>
-                    </li>
-                )
+    const { id, type } = props;
+    const { setGlobalState, filters: { 
+            activeTab, availabilityFilter
             }
-        </Context.Consumer>
-    )
+        } = useContext(Context);
```
Comment (noudadrichem) [inline components/Tabs/Tab.tsx:15]: Not sure what formatting is doing here.

File: components/Tabs/Tab.tsx [hunk 4]
```diff
@@ -1,32 +1,43 @@
...
+                            ? setGlobalState({ filters:{availabilityFilter, activeTab: 'all'} })
+                            : setGlobalState({ filters:{availabilityFilter, activeTab: id} })
+                        : isActive
+                        ? setGlobalState({
+                              filters:{ activeTab ,availabilityFilter: 'none'},
+                          })
+                        : setGlobalState({
+                              filters:{ activeTab, availabilityFilter: id},
+                          })
+                }
```
Comment (noudadrichem) [inline components/Tabs/Tab.tsx:32]: Same goes for this about logic, Maybe it's nicer to put this inside a more seperate function?

e.g. `function handleGBStatusFilter() {}`

File: components/Tabs/index.tsx [hunk 5]
```diff
@@ -8,33 +8,67 @@ import Select from '../Select';
...
     const state = useContext<InititalState>(Context);
 
     return (
-        <div className="filters">
-            <div className="tabs">
-                <label className="label">Choose your keycap type</label>
-                <ul>
-                    {state.tabs.map((tab: String, idx: number) => <Tab id={tab} key={idx} />)}
-                </ul>
+        <Fragment>
```
Comment (noudadrichem) [inline components/Tabs/index.tsx:15]: I used `<>` everywhere. Nicer to be consistent in that aswell.

File: pages/index.tsx [hunk 6]
```diff
@@ -30,63 +30,76 @@ interface HomeProps {
...
+        function handleTabChange() {
+            fetchMoreWhenSearched();
+        },
+        [state.filters.activeTab, state.filters.availabilityFilter]
+    );
+
+    useEffect(
+        function handleRefetchingOnBottomOfPage() {
+            const isEndReached =
+                state.keycapsets.length === state.allKeycapsetsCount;
```
Comment (noudadrichem) [inline pages/index.tsx:59]: The formatting here was also fine how it was? this is adding more lines and more unreadability for now reason imo.

File: components/Images.tsx (hunk 1 above)
Comment (noudadrichem) [reply components/Images.tsx:41]: We could also add an Status enum to the typings?

File: components/Tabs/Tab.tsx (hunk 4 above)
Comment (grantglidewell) [reply components/Tabs/Tab.tsx:32]: This is an insane nested ternary lol. I forgot to come back to refactor it

File: pages/index.tsx (hunk 6 above)
Comment (grantglidewell) [reply pages/index.tsx:59]: How is the prettiest integration going? Running that in this PR would resolve quite a few issues you’ve mentioned.

**Comment by grantglidewell** [comment]
@noudadrichem Ive addressed some of the issues you mentioned, If you want me to I have a package I wrote to automatically implement eslint, prettier, and pre-commit hooks. you can try it out by running `npx tag-standards` in the root directory of the repo. Some of the configuration might conflict with your preferences but it's a starting place. I was developed around Gatsby projects and did not take TS into account but might help jumpstart that effort.

let me know when the linting/prettier standards are in place and I can get the rest of this ready. You are also welcome to make the changes if you want, I wouldnt be offended. Really glad this is all moving forward and happy to contribute.

**Comment by noudadrichem** [comment]
I'm gonna check that out! Will work on misc stuff for KCS this weekend! Also re-look design for these filters because more may be added later on!

**Comment by noudadrichem** [comment]
As you can see I continued development in another branch to add some styling.
Wanted to let you know i'll put this online but will edit design asap.

**Comment by noudadrichem** [comment]
Do you want to implement the filters like you explained in a comment? With `allKeycapsets` and `filteredSets` @grantglidewell. If not I'll put this on a higher prio for myself.

**Comment by grantglidewell** [comment]
I can work on it this week if that timeframe works for you. Im glad this merged and it going out. Totally agree the styling should be different from what it is.

**Pull request merged**